File: chatbot/backend/src/llm/gemini.ts
import { LLMProvider } from './interface'
import { ChatMessage, LLMConfig } from '../config'

/**
 * Google Gemini provider — supports Gemini 1.5 Flash, Gemini 1.5 Pro, Gemini 2.0.
 * Uses the Gemini API's streaming endpoint.
 */
export class GeminiProvider implements LLMProvider {
  readonly name = 'Gemini'

  isConfigured(config: LLMConfig): boolean {
    return !!config.apiKey
  }

  async chat(
    messages: ChatMessage[],
    config: LLMConfig
  ): Promise<ReadableStream<string>> {
    if (!this.isConfigured(config)) {
      throw new Error('Gemini API key is not configured. Set GEMINI_API_KEY via `npx wrangler secret put GEMINI_API_KEY`')
    }

    // Convert our message format to Gemini's format
    const { systemInstruction, contents } = this.convertMessages(messages)

    const body: Record<string, unknown> = {
      contents,
      generationConfig: {
        maxOutputTokens: config.maxTokens,
        temperature: config.temperature,
      },
    }

    // Add system instruction if present
    if (systemInstruction) {
      body.systemInstruction = {
        parts: [{ text: systemInstruction }],
      }
    }

    const url = `https://generativelanguage.googleapis.com/v1beta/models/${config.model}:streamGenerateContent?alt=sse&key=${config.apiKey}`

    const response = await fetch(url, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify(body),
    })

    if (!response.ok) {
      const error = await response.text()
      throw new Error(`Gemini API error (${response.status}): ${error}`)
    }

    if (!response.body) {
      throw new Error('Gemini response body is null')
    }

    return parseGeminiStream(response.body as ReadableStream<Uint8Array>)
  }

  private convertMessages(messages: ChatMessage[]): {
    systemInstruction?: string
    contents: Array<{ role: string; parts: Array<{ text: string }> }>
  } {
    let systemInstruction: string | undefined
    const contents: Array<{
      role: string
      parts: Array<{ text: string }>
    }> = []

    for (const msg of messages) {
      if (msg.role === 'system') {
        systemInstruction = (systemInstruction || '') + msg.content + '\n'
      } else if (msg.role === 'user') {
        contents.push({ role: 'user', parts: [{ text: msg.content }] })
      } else if (msg.role === 'assistant') {
        contents.push({ role: 'model', parts: [{ text: msg.content }] })
      }
    }

    return { systemInstruction: systemInstruction?.trim(), contents }
  }
}

/**
 * Parse Gemini's SSE stream.
 * Gemini sends: data: {"candidates":[{"content":{"parts":[{"text":"content"}]}}]}
 */
async function parseGeminiStream(
  stream: ReadableStream<Uint8Array>
): Promise<ReadableStream<string>> {
  const decoder = new TextDecoder()
  const reader = stream.getReader()

  return new ReadableStream<string>({
    async start(controller) {
      let buffer = ''

      try {
        while (true) {
          const { done, value } = await reader.read()
          if (done) break

          buffer += decoder.decode(value, { stream: true })
          const lines = buffer.split('\n')
          buffer = lines.pop() || ''

          for (const line of lines) {
            const trimmed = line.trim()
            if (!trimmed || trimmed === '[DONE]') continue

            if (trimmed.startsWith('data: ')) {
              try {
                const json = JSON.parse(trimmed.slice(6))
                const text =
                  json.candidates?.[0]?.content?.parts?.[0]?.text || ''
                if (text) {
                  controller.enqueue(text)
                }
              } catch {
                // Skip malformed JSON
              }
            }
          }
        }
      } catch (err) {
        controller.error(err)
      } finally {
        controller.close()
        reader.releaseLock()
      }
    },
  })
}

File: chatbot/backend-python/src/llm/gemini.py
"""
Google Gemini provider — supports Gemini 1.5 Flash, Gemini 1.5 Pro, Gemini 2.0.

Uses the Gemini API directly via HTTP (no Python SDK dependency).
"""

import json
from typing import AsyncGenerator, List, Optional, Tuple

from .interface import LLMProvider, ChatMessage
from ..config import LLMConfig


class GeminiProvider(LLMProvider):
    def __init__(self, config: LLMConfig):
        self._config = config

    @property
    def name(self) -> str:
        return "Gemini"

    def is_configured(self) -> bool:
        return bool(self._config.api_key)

    async def chat(
        self,
        messages: List[ChatMessage],
    ) -> AsyncGenerator[str, None]:
        system_instruction, contents = self._convert_messages(messages)

        body: dict = {
            "contents": contents,
            "generationConfig": {
                "maxOutputTokens": self._config.max_tokens,
                "temperature": self._config.temperature,
            },
        }

        if system_instruction:
            body["systemInstruction"] = {"parts": [{"text": system_instruction}]}

        url = (
            f"https://generativelanguage.googleapis.com/v1beta/models/"
            f"{self._config.model}:streamGenerateContent"
            f"?alt=sse&key={self._config.api_key}"
        )

        import httpx

        async with httpx.AsyncClient() as client:
            async with client.stream(
                "POST", url, json=body, headers={"Content-Type": "application/json"}
            ) as response:
                response.raise_for_status()
                buffer = ""
                async for raw_line in response.aiter_lines():
                    buffer += raw_line
                    lines = buffer.split("\n")
                    buffer = lines.pop() if lines else ""

                    for line in lines:
                        line = line.strip()
                        if not line or line == "[DONE]":
                            continue
                        if line.startswith("data: "):
                            try:
                                data = json.loads(line[6:])
                                text = (
                                    data.get("candidates", [{}])[0]
                                    .get("content", {})
                                    .get("parts", [{}])[0]
                                    .get("text", "")
                                )
                                if text:
                                    yield text
                            except json.JSONDecodeError:
                                continue

    def _convert_messages(
        self, messages: List[ChatMessage]
    ) -> Tuple[Optional[str], List[dict]]:
        system_parts: list[str] = []
        contents: list[dict] = []

        for msg in messages:
            if msg.role == "system":
                system_parts.append(msg.content)
            elif msg.role == "user":
                contents.append({"role": "user", "parts": [{"text": msg.content}]})
            elif msg.role == "assistant":
                contents.append({"role": "model", "parts": [{"text": msg.content}]})

        system_instruction = "\n".join(system_parts).strip() or None
        return system_instruction, contents
